Update command injection docs (#741)
returntocorp/semgrep-docs: docs/cheat-sheets/javascript-code-injection.mdx

---
slug: javascript-code-injection
---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"

# Code injection prevention for JavaScript

<CodeInjectionIntro />

## 1. Executing or evaluating code

### 1.A. running code with `vm` module

> The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not a security mechanism. Do not use it to run untrusted code. If code passed to vm functions is controlled by user input it could result in command injection. 

Author: ```suggestion
The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not a security mechanism. Do not use it to run untrusted code. If code passed to vm functions is controlled by user input it could result in code injection. 
```

---
slug: javascript-code-injection
---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"

# Code injection prevention for JavaScript

<CodeInjectionIntro />

## 1. Executing or evaluating code

### 1.A. running code with `vm` module

The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not a security mechanism. Do not use it to run untrusted code. If code passed to vm functions is controlled by user input it could result in code injection. 

Example:

> ```javascript
> const vm = require('vm');
> const contextObject = { globalVar: 1 };
> 
> // safe
> vm.runInContext('globalVar *= 2;', contextObject);
> 
> // vulnerable 
> let userInput = 'this.constructor.constructor("return process.env")()'; // Value supplied by user input
> vm.runInContext(`globalVar = ${userInput};`, contextObject);
> ```
> 
> ```javascript
> const vm = require('vm');
> // safe
> const code = `return 'hello ' + name`
> vm.compileFunction(code, [], { parsingContext: vm.createContext({ name: 'name' }) })
> 
> // vulnerable
> let userInput = '1; while (true) {}'  // Value supplied by user input
> const code = `let x = ${userInput}; return 'hello ' + x`
> vm.compileFunction(code, [], { parsingContext: vm.createContext({ name: 'name' }) })
> ```

Author: ```suggestion
```javascript
const vm = require('vm');
const contextObject = { globalVar: 1 };

// safe
vm.runInContext('globalVar *= 2;', contextObject);

// vulnerable
let userInput = 'this.constructor.constructor("return process.env")()'; // Value supplied by user input
vm.runInContext(`globalVar = ${userInput};`, contextObject);

// safe
const code = `return 'hello ' + name`
vm.compileFunction(code, [], { parsingContext: vm.createContext({ name: 'name' }) })

// vulnerable
let userInput = '1; while (true)
```
```

#### References

- [`vm` module documentation](https://nodejs.org/dist/latest-v14.x/docs/api/vm.html)

#### Mitigation

> Do not let user input in `vm` functions.

Author: ```suggestion
- Don't use the `vm` module for running untrusted code.
- If you need to use the `vm` module functions with non-literal values, ensure that executed content is not controllable by external sources.
- If it’s not possible, strip everything except alphanumeric characters from the input.
```

---
slug: javascript-code-injection
---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"

# Code injection prevention for JavaScript

<CodeInjectionIntro />

#### Check your project for these conditions:

```bash
semgrep scan --config="r/javascript.lang.security.audit.vm-injection.vm-runincontext-context-injection" --config="r/javascript.lang.security.detect-eval-with-expression.detect-eval-with-expression"
```

---

## 1. Executing or evaluating code

### 1.A. running code with VM module

> The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not secure. Do not use it to run untrusted code. If code passed to `vm` functions is controlled by user input it could result in code injection. See [VM (executing JavaScript)](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html#vm-executing-javascript) documentation.

Reviewer: ```suggestion
The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not secure. Do not use it to run untrusted code. If user input is used as part of the code passed to `vm` functions, it could result in code injection. See [VM (executing JavaScript)](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html#vm-executing-javascript) documentation.
```

Author: ```suggestion
The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not secure. Do not use it to run untrusted code. If user input is used as a part of the code passed to `vm` functions, it can result in code injection. See [VM (executing JavaScript)](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html#vm-executing-javascript) documentation.
```

The following list documents all potentially vulnerable functions that compile and execute code from the `vm` module:
- `vm.runInContext`
- `vm.runInNewContext`
- `vm.runInThisContext`
- `vm.compileFunction`
- `new vm.Script`
- `new vm.SourceTextModule`

Example:

```javascript
const vm = require('vm');
const contextObject = { globalVar: 1 };

// safe
vm.runInContext('globalVar *= 2;', contextObject);

// vulnerable
let userInput = 'this.constructor.constructor("return process.env")()'; // Value supplied by user input
vm.runInContext(`globalVar = ${userInput};`, contextObject);

// safe
const code = `return 'hello ' + name`
vm.compileFunction(code, [], { parsingContext: vm.createContext({ name: 'name' }) })

// vulnerable
let userInput = '1; while (true)
```

#### References

- [VM module documentation](https://nodejs.org/dist/latest-v14.x/docs/api/vm.html)

#### Mitigation

- Don't use the `vm` module for running untrusted code.

> - If you need to use functions of the `vm` module with non-literal values, ensure that executed content is not controllable by external sources.

Reviewer: ```suggestion
- If you need to use functions of the `vm` module with non-literal values, ensure that the content being executed cannot be controlled by external sources.
```

- If it's not possible, strip everything except alphanumeric characters from the input.

#### Semgrep rule

<LinkToRegistryRule ruleId="javascript.lang.security.audit.vm-injection.vm-runincontext-context-injection" />

### 1.B. eval or new Function

The `eval()` or `new Function()` function evaluates JavaScript code represented as a string. Executing JavaScript from a string is an enormous security risk. It is far too easy for a bad actor to run arbitrary code when you use `eval()` or `new Function()`.

Example:

```javascript
const userInput = "1;require('child_process').exec('cat /etc/passwd')"
eval(`var x = ${userInput}`)

const userInput = "require('child_process').exec('cat /etc/passwd') && console.log"
var x = new Function(`return ${userInput}(a,b)`)
```

#### References

- [Never use eval()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!) in MDN Web Docs documentation.

#### Mitigation

- Don't use `eval()` or `new Function()` if possible.

> - If you need to use `eval()` or `new Function()` with non-literal values, ensure that executed content is not controllable by external sources.

Reviewer: ```suggestion
- If you need to use `eval()` or `new Function()` with non-literal values, ensure that the content being executed cannot be controlled by external sources.
```

Author: ```suggestion
- If you need to use `eval()` or `new Function()` with non-literal values, ensure that the executed content cannot be controlled by external sources.
```

---
slug: javascript-code-injection
---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"

# Code injection prevention for JavaScript

<CodeInjectionIntro />

#### Check your project for these conditions:

```bash
semgrep scan --config="r/javascript.lang.security.audit.vm-injection.vm-runincontext-context-injection" --config="r/javascript.lang.security.detect-eval-with-expression.detect-eval-with-expression"
```

---

## 1. Executing or evaluating code

### 1.A. running code with VM module

The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not secure. Do not use it to run untrusted code. If code passed to `vm` functions is controlled by user input it could result in code injection. 

:::caution

> Do not use the `vm` module to run untrusted code as this module is not a security mechanism. See [VM (executing JavaScript)](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html#vm-executing-javascript) documentation.

Reviewer 2: I think we can remove the callout. It basically repeats the first paragraph.

Also, "This module is not a security mechanism" - I'm not sure what the writer is trying to convey. the vm module can run code. It's not meant to _be_ a security mechanism. It's like innerHTML, it's insecure by virtue of its function and needs care when used.

Author: I agree, removing in https://github.com/returntocorp/semgrep-docs/pull/741/files#r1064635443

> The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not secure. Do not use it to run untrusted code. If code passed to `vm` functions is controlled by user input it could result in code injection. 
> 
> :::caution
> Do not use the `vm` module to run untrusted code as this module is not a security mechanism. See [VM (executing JavaScript)](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html#vm-executing-javascript) documentation.
> :::

Author: ```suggestion
The `vm` module enables compiling and running code within V8 Virtual Machine contexts. The `vm` module is not secure. Do not use it to run untrusted code. If code passed to `vm` functions is controlled by user input it could result in code injection. See [VM (executing JavaScript)](https://nodejs.org/dist/latest-v18.x/docs/api/vm.html#vm-executing-javascript) documentation.
```
